### ISSUE TITLE
Telegram Topics

### ISSUE THREAD

**Issue by FarisZR** [mautrix/telegram#865]
Telegram announced a new Topics feature in the latest update, some thing similar to Matrix Spaces.
https://telegram.org/blog/topics-in-groups-collectible-usernames

It would be great if the bridge would support it as separate group, so a topic would be treated as a new group.

So users could create a space on matrix and use Topics on Telegram to deliver the same experience.

**Comment by mrclauss** [comment]
+1

**Comment by kn0wmad** [comment]
Topics seems to actually break the bot for us, unfortunately.  It would be ideal if the bot could map room-to-room, then we could mirror Matrix/Telegram

**Comment by valerio-bozzolan** [comment]
I agree to handling a specific "Telegram group topic" to a specific Matrix room.

If developers need to do tests, feel free to ask me and I can invite you in a public Telegram group with topics enabled.

**Comment by joepie91** [comment]
From a glance, it seems like the mapping works something like this:

- If a message is a Matrix-side reply to another message, it is grouped under the Topic that the original message was in, as a Telegram-side reply to that original message
- If a message is a Matrix-side reply to the "pseudo-event" that represents the *creation* of the Topic, it is grouped under that Topic as a *non-reply* on the Telegram side (this is what all the replies to version errors are coming from, they're just topic-tagged messages)
- If a message is *not* a Matrix-side reply, it ends up in what seems like a "default" Topic

I do feel like a Space of rooms, one per Topic, would probably be the most viable approach here.

**Comment by FarisZR** [comment]
Is there a way to push this feature to get implemented?, a bounty of some sort maybe?

**Comment by git-bruh** [comment]
Gonna take a shot at it, stumbled upon some useful leads wrt topics and their interaction with the TG api

**Comment by git-bruh** [comment]
Hmm, turned out to be a bit more involved than I first thought. The API stuff with topics shouldn't be a huge issue, but integrating the logic of mapping topics to rooms seems to be a problematic task, since the whole codebase is modeled around the assumption of 1 TG group -> 1 Matrix room. Some of the issues I faced:

* The `portal` table https://github.com/mautrix/telegram/blob/master/mautrix_telegram/db/upgrade/v00_latest_revision.py#L34-L58 has a `UNIQUE` constraint on `tgid`, so that needs to be removed and another column for `topic_id` be added. But it adds some redundant data, eg `peer_type`, `tg_receiver` etc irrelevant for a topic since they're the same as the "parent" group. Creating a separate table would be unmaintainable since each query would have to be special cased.

* Bans / kicks would be an issue aswell since you'd have to make them reflect across each room corresponding to topics in a TG group

* Some stuff would need to be special cased to not run for rooms that are topics, eg. backfilling

* A conditional for `topic_id` needs to be spammed everywhere where data from the telegram API is involved

So far I only have the logic for getting the topic ID from a TG message https://github.com/git-bruh/telegram/commit/1dc0fc6ad96cfe8cc820f81710e035c3063d8e30

**Comment by ashledombos** [comment]
Is it possible to somehow associate a topic on telegram side, to a thread on Matrix side? This feature seems closer to topics than Space/rooms.

**Comment by git-bruh** [comment]
I did think of that aswell as it would save a ton of hassle but the UX didn't feel right, like an empty room with just N threads. Probably better to do it this way though, makes it possible to avoid handling membership event for every topic room

**Comment by KF-Art** [comment]
So for now it not possible to implement cleanly bridge a group which has topics? 

For now the only workaround I could think about was to use threads which each one corresponds to one topic. Of course this is not practical, because I would have to tell users to use only those threads and not replying outside them, but also it makes that all messages sent from Matrix in a Topic would be marked as reply of the first message.

I know that these things are complicated but it would be great that the bridge supports topics in the future, regardless is through spaces+rooms or threads.

**Comment by ManDay** [comment]
Personally, I wouldn't want a fixed correspondence between a dynamic number of topics and threads, but rather the ability to create **one bridge** to a specific topic within a TG group. While Topics and Threads seem to be a matching analogy, I'm coming from a case where the desired mapping is Topic to Room, and we wouldn't want multiple Topics (or Threads) in the room.

**Comment by ManDay** [comment]
What about the other direction, though? The bridge can only be useful if it works in both.

**Comment by valerio-bozzolan** [comment]
Well, I guess:

- if we are replying to a message that starts with `[XXX]` try sending to the `XXX` Channel
- if we are not replying to anything but it starts with that prefix, try sending there
- send to the main channel otherwise

**Comment by TomTheDragon** [comment]
It would be great if we had the support for topics, since more and more groups switch over to just one single one with topics.
We recently switched over from Matterbridge to Mautrix-Telegram for most of our groups, since it integrates so much better between Matrix and Telegram. But we also have one group that we cant bridge, since it uses topics.

I am not sure if this helps but thats the commit that gave Matterbridge its topics support
https://github.com/42wim/matterbridge/commit/601f48a50ea31a9631011766c680ecef931ddafc

**Comment by ayoahha** [comment]
Is there any update on this feature ?
Matching Telegram topics with rooms

**Comment by Gredin67** [comment]
@git-bruh 
> So far I only have the logic for getting the topic ID from a TG message https://github.com/git-bruh/telegram/commit/1dc0fc6ad96cfe8cc820f81710e035c3063d8e30

Could you use that to implement that ? :

> puts the topic as prefix of the message body.

**Comment by Gredin67** [comment]
> * The `portal` table https://github.com/mautrix/telegram/blob/master/mautrix_telegram/db/upgrade/v00_latest_revision.py#L34-L58 has a `UNIQUE` constraint on `tgid`, so that needs to be removed and another column for `topic_id` be added. But it adds some redundant data, eg `peer_type`, `tg_receiver` etc irrelevant for a topic since they're the same as the "parent" group. Creating a separate table would be unmaintainable since each query would have to be special cased.
> 
> * Bans / kicks would be an issue aswell since you'd have to make them reflect across each room corresponding to topics in a TG group

@git-bruh as far as I know a Matrix Space is nothing else than a room containing rooms. Wouldn't the solution be to modify the Matrix Spec to allow binding rooms state (state_events/permissions/participants) to that of the Space it belongs to ? 
* Bridge `state events` from the telegram group to the space
* Bridge `message events` to one room per telegram group topic

see. https://spec.matrix.org/latest/#room-structure

**Comment by Gredin67** [comment]
We should have a canonical space, from which the "topic rooms" would inherit their state.
> m.space.parent events can additionally include a canonical boolean key in their content to denote that the parent space is the main/primary space for the room.
https://spec.matrix.org/v1.9/client-server-api/#mspaceparent-relationships

**Comment by Facni** [comment]
Any update?

**Comment by NepNep21** [comment]
Wouldn't it make more sense to bridge topics as threads rather than rooms in a space?

**Comment by maltee1** [comment]
Both approaches have their pros and cons UX wise, but bridging as topics is probably a lot less effort.

**Comment by zertap** [comment]
More people have started using Telegram Topics, and currently the bridge is unusable for me without support for topics.
I've resigned to using matterbridge for these instead.

**Comment by c-holtermann** [comment]
> So far I only have the logic for getting the topic ID from a TG message https://github.com/git-bruh/telegram/commit/1dc0fc6ad96cfe8cc820f81710e035c3063d8e30

@git-bruh do you still have the code? The link to your repository does not exist anymore.
